zoom_map.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/python3
+
+import numpy
+import PIL.Image
+import PIL.ImageTk
+
+
+class ZoomMap:
+    def __init__(self, matrix):
+        self._matrix = matrix
+        self._zoom_level = 0
+        self._pyramid = []
+        for _ in range(5):  # TODO: replace this with a real loop
+            self._pyramid.append(self._to_image(matrix))
+            # Combine 2x2 squares of pixels to make the next level.
+            nr, nc = [(value // 2) * 2 for value in matrix.shape]
+            quads = [matrix[row:nr:2, col:nc:2]
+                     for row in [0, 1] for col in [0, 1]]
+            # We want the following outcomes when combining a 2x2 square into a
+            # single pixel:
+            #   - If none of the 4 pixels is set, we should not be set.
+            #   - If 1 of the 4 pixels is set, we're set a quarter of the time.
+            #   - If 2 of the 4 pixels are set, we're set half the time.
+            #   - It's impossible to have 3 of the 4 pixels set.
+            #   - If all 4 pixels are set, this one should be set, too.
+            # To discuss the times when half the pixels are set:
+            #   - If the two that are set are on the main diagonal, we should be
+            #     set. It's good to make diagonals easy to see.
+            #   - If the two that are set are off the main diagonal, we should
+            #     not be set.
+            #   - If the two that are set are adjacent to each other, we should
+            #     be set half the time.
+            # To discuss times when 1 pixel is set:
+            #   - If the 1 pixel is off the diagonal, it might be part of a
+            #     large diagonal line shifted 1 pixel off of our diagonal. Half
+            #     of these should be set.
+            #   - If the 1 pixel is on the diagonal, we should not be set (so
+            #     that we're set a quarter of the time overall).
+            # To satisfy all these conditions, we should be set either if both
+            # pixels on the diagonal are set or if 1 pixel off the diagonal is
+            # set.
+            matrix = ((quads[0] & quads[3]) |
+                      (quads[1] & numpy.logical_not(quads[2])))
```

Comment:
Just for the sake of writing it down: I wonder if this could be made  more efficient (and smaller codewise) using something like resizing with bicubic interpoliation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.imresize.html